File: remove_n_element.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Couldn't solve before class


// Your code here along with comments explaining your approach
//initialize dummyHead pointing to head.next
//keep first pointer at head and increment until n>0
//keep second pointer at dummyHead so that the difference between first and second pointer is n
//increment both first and second until first reaches null
//after that delete the node and increment second by 2
//return dummy node . next

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
      
        if(head == null) return head;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        
        ListNode first = head, second = dummy;
        
        int count = n;
        while(count > 0){
            first = first.next;
            count--;
        }
        
        while(first!=null){
            first = first.next;
            second = second.next;
        }
        
        second.next = second.next.next;
        return dummy.next;
        
    }
}

File: remove_nelement.py
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        count = 1
        prev, cur, ptr = None, None, head
        while(ptr):
            if count == n:
                cur = head
            elif count > n:
                prev = cur
                cur = cur.next
                
            ptr = ptr.next
            count += 1
            
        if cur is None:
            return head
        if prev is None:
            return head.next
        
        prev.next = cur.next
        return head